libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -335,278 +343,566 @@ void AC_PosControl::set_dt(float delta_sec)
-    if (is_positive(speed_up) && is_negative(speed_down)) {
-        _speed_down_cms = speed_down;
-        _speed_up_cms = speed_up;
-        _flags.recalc_leash_z = true;
-        calc_leash_length_z();
+    // adjust desired alt if motors have not hit their limits
+    update_pos_vel_accel_z(_pos_target, _vel_desired, _accel_desired, _dt,
+        _motors.limit.throttle_lower, _motors.limit.throttle_upper,
+        _p_pos_z.get_error(), _pid_vel_z.get_error());
+
+    float vel_max_xy_cms = _vel_max_xy_cms;
+    float vel_max_z_cms = 0.0f;
+    if ( is_positive(dest_vector.length_squared()) ) {
+        dest_vector.normalize();
+        float dest_vector_xy_length = Vector2f{dest_vector.x, dest_vector.y}.length();
+
+        float vel_max_cms = kinematic_limit(dest_vector, _vel_max_xy_cms, _vel_max_up_cms, _vel_max_down_cms);
+        vel_max_xy_cms = vel_max_cms * dest_vector_xy_length;
+        vel_max_z_cms = vel_max_cms * dest_vector.z;
     }
+
+
+    Vector3f vel;
+    Vector3f accel;
+    shape_pos_vel_accel_xy(pos, vel, accel, _pos_target, _vel_desired, _accel_desired,
+        vel_max_xy_cms, _vel_max_xy_cms, _accel_max_xy_cmss, _tc_xy_s, _dt);
+    shape_pos_vel_accel_z(pos, vel, accel,
+        _pos_target, _vel_desired, _accel_desired,
+        vel_max_z_cms, _vel_max_down_cms, _vel_max_up_cms,
+        -constrain_float(accel_z_cmss, 0.0f, 750.0f), accel_z_cmss,
+        _tc_z_s, _dt);
 }
 
-/// set_max_accel_z - set the maximum vertical acceleration in cm/s/s
-void AC_PosControl::set_max_accel_z(float accel_cmss)
+
+///
+/// Lateral position controller
+///
+
+/// set_max_speed_accel_xy - set the maximum horizontal speed in cm/s and acceleration in cm/s/s
+void AC_PosControl::set_max_speed_accel_xy(float speed_cms, float accel_cmss, float accel_limit_cmss)
 {
-    // exit immediately if no change in acceleration
-    if (is_equal(_accel_z_cms, accel_cmss)) {
+    // return immediately if no change
+    if (is_equal(_vel_max_xy_cms, speed_cms) && is_equal(_accel_max_xy_cmss, accel_cmss) && is_equal(_accel_limit_xy_cmss, accel_limit_cmss)) {
         return;
     }
+    _vel_max_xy_cms = speed_cms;
+    _accel_max_xy_cmss = accel_cmss;
+    _accel_limit_xy_cmss = accel_limit_cmss;
 
-    _accel_z_cms = accel_cmss;
-    _flags.recalc_leash_z = true;
-    calc_leash_length_z();
+    if ( is_positive(accel_limit_cmss)) {
+        // Use half the maximum acceleration for the position controller approach limit to ensure velocity controller has sufficient head room to operate effectively.
+        accel_cmss = MIN(_accel_max_xy_cmss, 0.5f * accel_limit_cmss);
+    }
+
+    _p_pos_xy.set_limits(0.0f, _vel_max_xy_cms, accel_cmss, 0.0f);
+
+    // ensure the horizontal time constant is not less than the vehicle is capable of
+    float lean_angle = _accel_max_xy_cmss / (GRAVITY_MSS * 100.0 * M_PI / 18000.0);
+    float angle_accel = MIN(_attitude_control.get_accel_pitch_max(), _attitude_control.get_accel_roll_max());
```

Comment:
we probably need to protect against divide-by-zero here because I think _attitude_control.get_accel_pitch_max() or _attitude_control.get_accel_roll_max() could return zero.